File: scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/PPMLContext.scala
/*
 * Copyright 2016 The BigDL Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.analytics.bigdl.ppml

import com.intel.analytics.bigdl.dllib.NNContext.{checkScalaVersion, checkSparkVersion, createSparkConf, initConf, initNNContext}
import com.intel.analytics.bigdl.dllib.utils.Log4Error
import com.intel.analytics.bigdl.ppml.crypto.{AES_CBC_PKCS5PADDING, BigDLEncrypt, Crypto, CryptoMode, DECRYPT, ENCRYPT, EncryptRuntimeException, PLAIN_TEXT}
import com.intel.analytics.bigdl.ppml.utils.Supportive
import com.intel.analytics.bigdl.ppml.kms.common.{KeyLoader, KeyLoaderManagement}
import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.input.PortableDataStream
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.{DataFrame, DataFrameReader, DataFrameWriter, Row, SparkSession}
import com.intel.analytics.bigdl.ppml.kms.{AzureKeyManagementService, EHSMKeyManagementService, KMS_CONVENTION,
KeyManagementService, SimpleKeyManagementService, BigDLKeyManagementService}
import com.intel.analytics.bigdl.ppml.crypto.dataframe.{EncryptedDataFrameReader, EncryptedDataFrameWriter}
import org.apache.hadoop.fs.Path

/**
 * PPMLContext who wraps a SparkSession and provides read functions to
 * read encrypted data files to plain-text RDD or DataFrame, also provides
 * write functions to save DataFrame to encrypted data files.
 */
class PPMLContext {
  protected val keyLoaderManagement = new KeyLoaderManagement
  protected var sparkSession: SparkSession = null
  protected var defaultKey: String = ""

  /**
   * Read data files into RDD[String]
   * @param path data file path
   * @param minPartitions min partitions
   * @param cryptoMode crypto mode, such as PLAIN_TEXT or AES_CBC_PKCS5PADDING
   * @param kmsName for multi-data source/KMS mode, name of kms which this data source uses
   * @param primaryKeyName use the name-specific primary key to decrypt data key in file meata
   * @return
   */
  def textFile(path: String,
               minPartitions: Int = sparkSession.sparkContext.defaultMinPartitions,
               cryptoMode: CryptoMode = PLAIN_TEXT,
               primaryKeyName: String = "defaultKey"): RDD[String] = {
    cryptoMode match {
      case PLAIN_TEXT =>
        sparkSession.sparkContext.textFile(path, minPartitions)
      case _ =>
        val dataKeyPlainText = primaryKeyName match {
          case "" =>
            keyLoaderManagement.retrieveKeyLoader(defaultKey)
                               .retrieveDataKeyPlainText(path)
          case _ =>
            keyLoaderManagement.retrieveKeyLoader(primaryKeyName)
                               .retrieveDataKeyPlainText(path)
        }
        PPMLContext.textFile(sparkSession.sparkContext, path, dataKeyPlainText,
                             cryptoMode, minPartitions)
    }
  }

  /**
   * Interface for loading data in external storage to Dataset.
   * @param cryptoMode crypto mode, such as PLAIN_TEXT or AES_CBC_PKCS5PADDING
   * @param primaryKeyName use the name-specific primary key to decrypt data key in file meata
   * @return a EncryptedDataFrameReader
   */
  def read(cryptoMode: CryptoMode,
           primaryKeyName: String = ""): EncryptedDataFrameReader = {
      primaryKeyName match {
        case "" =>
          new EncryptedDataFrameReader(sparkSession, cryptoMode,
            defaultKey, keyLoaderManagement)
        case _ =>
          new EncryptedDataFrameReader(sparkSession, cryptoMode,
            primaryKeyName, keyLoaderManagement)
      }
  }

  /**
   * Interface for saving the content of the non-streaming Dataset out into external storage.
   * @param dataFrame dataframe to save.
   * @param cryptoMode crypto mode, such as PLAIN_TEXT or AES_CBC_PKCS5PADDING
   * @param primaryKeyName use the name-specific primary key to decrypt data key in file meata
   * @return a DataFrameWriter[Row]
   */
  def write(dataFrame: DataFrame,
            cryptoMode: CryptoMode,
            primaryKeyName: String = ""): EncryptedDataFrameWriter = {
      primaryKeyName match {
        case "" =>
          new EncryptedDataFrameWriter(sparkSession, dataFrame, cryptoMode,
            defaultKey, keyLoaderManagement)
        case _ =>
          new EncryptedDataFrameWriter(sparkSession, dataFrame, cryptoMode,
            primaryKeyName, keyLoaderManagement)
      }
  }

  /**
   * Get SparkSession from PPMLContext
   * @return SparkSession in PPMLContext
   */
  def getSparkSession(): SparkSession = {
    sparkSession
  }
}


object PPMLContext{
  private[bigdl] def registerUDF(
        spark: SparkSession,
        cryptoMode: CryptoMode,
        dataKeyPlaintext: String) = {
    val bcKey = spark.sparkContext.broadcast(dataKeyPlaintext)
    val convertCase = (x: String) => {
      val crypto = Crypto(cryptoMode)
      crypto.init(cryptoMode, ENCRYPT, dataKeyPlaintext)
      new String(crypto.doFinal(x.getBytes)._1)
    }
    spark.udf.register("convertUDF", convertCase)
  }

  private[bigdl] def textFile(sc: SparkContext,
               path: String,
               dataKeyPlaintext: String,
               cryptoMode: CryptoMode,
               minPartitions: Int = -1): RDD[String] = {
    Log4Error.invalidInputError(dataKeyPlaintext != "",
      "dataKeyPlainText should not be empty, please loadKeys first.")
    val data: RDD[(String, PortableDataStream)] = if (minPartitions > 0) {
      sc.binaryFiles(path, minPartitions)
    } else {
      sc.binaryFiles(path)
    }
    data.mapPartitions { iterator => {
      Supportive.logger.info("Decrypting bytes with JavaAESCBC...")
      iterator.flatMap{dataStream =>
        val inputDataStream = dataStream._2.open()
        val crypto = Crypto(cryptoMode)
        val (encryptedDataKey, initializationVector) = crypto.getHeader(inputDataStream)
        crypto.init(cryptoMode, DECRYPT, dataKeyPlaintext)
        crypto.verifyHeader(initializationVector)
        crypto.decryptBigContent(inputDataStream)
      }
    }}
  }

  private[bigdl] def write(
        sparkSession: SparkSession,
        cryptoMode: CryptoMode,
        dataKeyPlaintext: String,
        dataFrame: DataFrame): DataFrameWriter[Row] = {
    val tableName = "ppml_save_table"
    dataFrame.createOrReplaceTempView(tableName)
    PPMLContext.registerUDF(sparkSession, cryptoMode, dataKeyPlaintext)
    // Select all and encrypt columns.
    val convertSql = "select " + dataFrame.schema.map(column =>
      "convertUDF(" + column.name + ") as " + column.name).mkString(", ") +
      " from " + tableName
    val df = sparkSession.sql(convertSql)
    df.write
  }

  def initPPMLContext(appName: String): PPMLContext = {
    initPPMLContext(null, appName)
  }

  def initPPMLContext(conf: SparkConf): PPMLContext = {
    initPPMLContext(conf)
  }

   /**
   * init ppml context with an existed SparkSession
   * @param sparkSession a SparkSession
   * @return a PPMLContext
   */
  def initPPMLContext(sparkSession: SparkSession): PPMLContext = {
    val conf = sparkSession.sparkContext.getConf
    Log4Error.invalidInputError(conf.contains("spark.hadoop.io.compression.codecs"),
        "spark.hadoop.io.compression.codecs not found!" +
        "If you want to init PPMLContext with an existing SparkSession, " +
        "must set the property before creating SparkSession!")
    Log4Error.invalidInputError(
      conf.get("spark.hadoop.io.compression.codecs")
      == "com.intel.analytics.bigdl.ppml.crypto.CryptoCodec",
      "If you want to init PPMLContext with an existing SparkSession, " +
      "spark.hadoop.io.compression.codecs property must be set to" +
      "com.intel.analytics.bigdl.ppml.crypto.CryptoCodec" +
      " before creating SparkSession!")
    val ppmlSc = loadPPMLContext(sparkSession)
    ppmlSc
  }

  /**
   * init ppml context with app name and ppml args
   * @param appName the name of this Application
   * @param ppmlArgs ppml arguments in a Map
   * @return a PPMLContext
   */
  def initPPMLContext(
        appName: String,
        ppmlArgs: Map[String, String]): PPMLContext = {
    initPPMLContext(null, appName, ppmlArgs)
  }

  /**
   * init ppml context with app name, SparkConf and ppml args
   * @param sparkConf a SparkConf
   * @param appName the name of this Application
   * @param ppmlArgs ppml arguments in a Map
   * @return a PPMLContext
   */
  def initPPMLContext(
        sparkConf: SparkConf,
        appName: String,
        ppmlArgs: Map[String, String]): PPMLContext = {
    val conf = createSparkConf(sparkConf)
    ppmlArgs.foreach { arg =>
      conf.set(arg._1, arg._2)
    }
    initPPMLContext(conf, appName)
  }

  /**
   * init ppml context with app name, SparkConf
   * @param sparkConf a SparkConf, ppml arguments are passed by this sparkconf.
   * @param appName the name of this Application
   * @return a PPMLContext
   */
  def initPPMLContext(sparkConf: SparkConf, appName: String): PPMLContext = {
    val conf = createSparkConf(sparkConf)
    conf.set("spark.hadoop.io.compression.codecs",
        "com.intel.analytics.bigdl.ppml.crypto.CryptoCodec")
    val sc = initNNContext(conf, appName)
    val sparkSession: SparkSession = SparkSession.builder().config(conf).getOrCreate()
    val ppmlSc = loadPPMLContext(sparkSession)
    ppmlSc
  }

  /**
   * load ppml context from spark session juding whether multiple KMSs exist
   * @param sparkSession a SparkSession
   * @return a PPMLContext
   */
  def loadPPMLContext(sparkSession: SparkSession): PPMLContext = {
    val ppmlSc = new PPMLContext
    ppmlSc.sparkSession = sparkSession
    val conf = sparkSession.sparkContext.getConf
    sparkSession.sparkContext.hadoopConfiguration.set(
      "spark.bigdl.encryter.type",
      conf.get("spark.bigdl.encryter.type", BigDLEncrypt.COMMON))
    val primaryKeyNames = getPrimaryKeyNames(conf)
    primaryKeyNames.foreach{
      primaryKeyName => {
        if (conf.contains(s"spark.bigdl.primaryKey.$primaryKeyName.plainText")) {
          val primaryKeyPlainText = conf.get(
            s"spark.bigdl.primaryKey.$primaryKeyName.plainText")
          ppmlSc.keyLoaderManagement
                .addKeyLoader(primaryKeyName,
                              KeyLoader(false, "", null, primaryKeyPlainText))
        } else {
          Log4Error.invalidInputError(
            conf.contains(s"spark.bigdl.primaryKey.$primaryKeyName.material"),
                          s"spark.bigdl.primaryKey.$primaryKeyName.material not found.")
            val primaryKeyMaterial = conf.get(
              s"spark.bigdl.primaryKey.$primaryKeyName.material")
            val kms = loadKmsOfPrimaryKey(conf, primaryKeyName)
            ppmlSc.keyLoaderManagement
                  .addKeyLoader(primaryKeyName,
                                KeyLoader(true, primaryKeyMaterial, kms, ""))
         }
      }
    }
    if (ppmlSc.keyLoaderManagement.count == 1) ppmlSc.defaultKey = primaryKeyNames(0)
    ppmlSc
  }

  def getPrimaryKeyNames(conf: SparkConf): Array[String] = {
    val prefix = "spark.bigdl.primaryKey"
    val properties: Array[Tuple2[String, String]] = conf.getAllWithPrefix(prefix)
    val names = for { v <- properties } yield v._1.split('.')(1)
    names.distinct
  }

  def loadKmsOfPrimaryKey(conf: SparkConf, primaryKeyName: String): KeyManagementService = {
    val kmsType = conf.get(s"spark.bigdl.primaryKey.$primaryKeyName.kms.type",
      defaultValue = KMS_CONVENTION.MODE_SIMPLE_KMS)
    val kms = kmsType match {
      case KMS_CONVENTION.MODE_EHSM_KMS =>
        val ip = conf.get(s"spark.bigdl.primaryKey.$primaryKeyName.kms.ip")
        val port = conf.get(s"spark.bigdl.primaryKey.$primaryKeyName.kms.port")
        val appId = conf.get(s"spark.bigdl.primaryKey.$primaryKeyName.kms.appId")
        val apiKey = conf.get(s"spark.bigdl.primaryKey.$primaryKeyName.kms.apiKey")
        new EHSMKeyManagementService(ip, port, appId, apiKey)
      case KMS_CONVENTION.MODE_SIMPLE_KMS =>
        val appId = conf.get(s"spark.bigdl.primaryKey.$primaryKeyName.kms.appId",
                             defaultValue = "simpleAPPID")
        val apiKey = conf.get(s"spark.bigdl.primaryKey.$primaryKeyName.kms.apiKey",
                              defaultValue = "simpleAPIKEY")
        SimpleKeyManagementService(appId, apiKey)
      case KMS_CONVENTION.MODE_AZURE_KMS =>
        val vaultName = conf.get(s"spark.bigdl.primaryKey.$primaryKeyName.kms.vault")
        val clientId = conf.get(s"spark.bigdl.primaryKey.$primaryKeyName.kms.clientId")
        new AzureKeyManagementService(vaultName, clientId)
      case KMS_CONVENTION.MODE_BIGDL_KMS =>
        val ip = conf.get(s"spark.bigdl.primaryKey.$primaryKeyName.kms.ip")
        val port = conf.get(s"spark.bigdl.primaryKey.$primaryKeyName.kms.port")
        val userName = conf.get(s"spark.bigdl.primaryKey.$primaryKeyName.kms.user")
        val userToken = conf.get(s"spark.bigdl.primaryKey.$primaryKeyName.kms.token")
        new BigDLKeyManagementService(ip, port, userName, userToken)
      case _ =>
        throw new EncryptRuntimeException("Wrong kms type")
    }
    kms
  }
}

File: python/ppml/src/bigdl/ppml/ppml_context.py
#
# Copyright 2016 The BigDL Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from bigdl.ppml.api import *
from bigdl.ppml.utils.log4Error import *
from enum import Enum

from pyspark.sql import SparkSession
from pyspark.ml.wrapper import JavaParams

def check(ppml_args, arg_name):
    try:
        value = ppml_args[arg_name]
        return value
    except KeyError:
        invalidInputError(False, "need argument " + arg_name)


class PPMLContext(JavaValue):
    def __init__(self, app_name, ppml_args=None, spark_conf=None):
        self.bigdl_type = "float"
        conf = {"spark.app.name": app_name,
                "spark.hadoop.io.compression.codecs": "com.intel.analytics.bigdl.ppml.crypto.CryptoCodec"}
        if spark_conf:
            for (k, v) in spark_conf.getAll():
                conf[k] = v
        if ppml_args:
            kms_type = ppml_args.get("kms_type", "")
            conf["spark.bigdl.primaryKey.defaultKey.kms.type"] = kms_type
            if kms_type == "SimpleKeyManagementService":
                conf["spark.bigdl.primaryKey.defaultKey.kms.appId"] = check(ppml_args, "app_id")
                conf["spark.bigdl.primaryKey.defaultKey.kms.apiKey"] = check(ppml_args, "api_key")
                conf["spark.bigdl.primaryKey.defaultKey.material"] = check(ppml_args, "primary_key_material")
            elif kms_type == "EHSMKeyManagementService":
                conf["spark.bigdl.primaryKey.defaultKey.kms.ip"] = check(ppml_args, "kms_server_ip")
                conf["spark.bigdl.primaryKey.defaultKey.kms.port"] = check(ppml_args, "kms_server_port")
                conf["spark.bigdl.primaryKey.defaultKey.kms.id"] = check(ppml_args, "app_id")
                conf["spark.bigdl.primaryKey.defaultKey.kms.apiKey"] = check(ppml_args, "api_key")
                conf["spark.bigdl.primaryKey.defaultKey.material"] = check(ppml_args, "primary_key_material")
            elif kms_type == "AzureKeyManagementService":
                conf["spark.bigdl.primaryKey.defaultKey.kms.vault"] = check(ppml_args, "vault")
                conf["spark.bigdl.primaryKey.defaultKey.kms.clientId"] = ppml_args.get("client_id", "")
                conf["spark.bigdl.primaryKey.defaultKey.material"] = check(ppml_args, "primary_key_material")
            elif kms_type == "BigDLKeyManagementService":
                conf["spark.bigdl.primaryKey.defaultKey.kms.ip"] = check(ppml_args, "kms_server_ip")
                conf["spark.bigdl.primaryKey.defaultKey.kms.port"] = check(ppml_args, "kms_server_port")
                conf["spark.bigdl.primaryKey.defaultKey.kms.user"] = check(ppml_args, "kms_user_name")
                conf["spark.bigdl.primaryKey.defaultKey.kms.token"] = check(ppml_args, "kms_user_token")
                conf["spark.bigdl.primaryKey.defaultKey.material"] = check(ppml_args, "primary_key_material")
            elif kms_type == "":
                conf["spark.bigdl.primaryKey.defaultKey.plainText"] = check(ppml_args, "primary_key_plaintext")
            else:
                invalidInputError(False, "invalid KMS type.")

        spark_conf = init_spark_conf(conf)

        sc = SparkContext.getOrCreate(spark_conf)

        self.spark = SparkSession.builder.getOrCreate()
        args = [self.spark._jsparkSession]
        super().__init__(None, self.bigdl_type, *args)

    def read(self, crypto_mode, primary_key_name = ""):
        if isinstance(crypto_mode, CryptoMode):
            crypto_mode = crypto_mode.value
        df_reader = callBigDlFunc(self.bigdl_type, "read",
                                  self.value, crypto_mode, primary_key_name)
        return EncryptedDataFrameReader(self.bigdl_type, df_reader)

    def write(self, dataframe, crypto_mode, primary_key_name = ""):
        if isinstance(crypto_mode, CryptoMode):
            crypto_mode = crypto_mode.value
        df_writer = callBigDlFunc(self.bigdl_type, "write", self.value,
                                  dataframe, crypto_mode, primary_key_name)
        return EncryptedDataFrameWriter(self.bigdl_type, df_writer)

    def textfile(self, path, min_partitions=None,
                 crypto_mode = "plain_text", primary_key_name = ""):
        if min_partitions is None:
            min_partitions = self.spark.sparkContext.defaultMinPartitions
        if isinstance(crypto_mode, CryptoMode):
            crypto_mode = crypto_mode.value
        return callBigDlFunc(self.bigdl_type, "textFile",
                             self.value, path, min_partitions,
                             crypto_mode, primary_key_name)

    def sql(self, sqlText):
        return callBigDlFunc(self.bigdl_type, "sql",
                             self.value, sqlText)

    def saveLightGBMModel(self, lightgbm_model, path,
            crypto_mode = "plain_text", primary_key_name = ""):
        if isinstance(crypto_mode, CryptoMode):
            crypto_mode = crypto_mode.value
        return callBigDlFunc(self.bigdl_type, "saveLightGBMModel",
                self.value, lightgbm_model._java_obj,
                path, crypto_mode, primary_key_name)

    def loadLightGBMClassificationModel(self, model_path,
            crypto_mode = "plain_text", primary_key_name = ""):
        if isinstance(crypto_mode, CryptoMode):
            crypto_mode = crypto_mode.value
        java_model = callBigDlFunc(self.bigdl_type,
                "loadLightGBMClassificationModel",
                self.value, model_path,
                crypto_mode, primary_key_name)
        return JavaParams._from_java(java_model)

    def loadLightGBMRegressionModel(self, model_path,
            crypto_mode = "plain_text", primary_key_name = ""):
        if isinstance(crypto_mode, CryptoMode):
            crypto_mode = crypto_mode.value
        java_model = callBigDLFunc(self.bigdl_type,
                "loadLightGBMRegressionModel", model_path,
                crypto_mode, primary_key_name)
        return JavaParams._from_java(java_model)

    def loadLightGBMRankerModel(self, model_path,
            crypto_mode = "plain_text", primary_key_name = ""):
        if isinstance(crypto_mode, CryptoMode):
            crypto_mode = crypto_mode.value
        java_model = callBigDLFunc(self.bigdl_type,
                "loadLightGBMRankerModel", model_path,
                crypto_mode, primary_key_name)
        return JavaParams._from_java(java_model)

class EncryptedDataFrameReader:
    def __init__(self, bigdl_type, df_reader):
        self.bigdl_type = bigdl_type
        self.df_reader = df_reader

    def schema(self, value):
        self.df_reader = callBigDlFunc(self.bigdl_type, "schema", self.df_reader, value)
        return self

    def option(self, key, value):
        self.df_reader = callBigDlFunc(self.bigdl_type, "option", self.df_reader, key, value)
        return self

    def csv(self, path):
        return callBigDlFunc(self.bigdl_type, "csv", self.df_reader, path)

    def parquet(self, path):
        return callBigDlFunc(self.bigdl_type, "parquet", self.df_reader, path)

    def json(self, path):
        return callBigDlFunc(self.bigdl_type, "json", self.df_reader, path)


class EncryptedDataFrameWriter:
    support_mode = {"overwrite", "append", "ignore", "error", "errorifexists"}

    def __init__(self, bigdl_type, df_writer):
        self.bigdl_type = bigdl_type
        self.df_writer = df_writer

    def option(self, key, value):
        self.df_writer = callBigDlFunc(self.bigdl_type, "option", self.df_writer, key, value)
        return self

    def mode(self, mode):
        invalidInputError(mode in EncryptedDataFrameWriter.support_mode,
                          "Unknown save mode: " + mode + "." +
                          "Accepted save modes are 'overwrite', 'append', 'ignore', 'error', 'errorifexists'.")
        self.df_writer = callBigDlFunc(self.bigdl_type, "mode", self.df_writer, mode)
        return self

    def csv(self, path):
        return callBigDlFunc(self.bigdl_type, "csv", self.df_writer, path)

    def parquet(self, path):
        return callBigDlFunc(self.bigdl_type, "parquet", self.df_writer, path)

    def json(self, path):
        return callBigDlFunc(self.bigdl_type, "json", self.df_writer, path)


class CryptoMode(Enum):
    """
    BigDL crypto mode for encrypt and decrypt data.
    """

    # CryptoMode PLAIN_TEXT
    PLAIN_TEXT = "plain_text"

    # CryptoMode AES_CBC_PKCS5PADDING
    AES_CBC_PKCS5PADDING = "AES/CBC/PKCS5Padding"

    # CryptoMode AES_GCM_V1 for parquet only
    AES_GCM_V1 = "AES_GCM_V1"

    # CryptoMode AES_GCM_CTR_V1 for parquet only
    AES_GCM_CTR_V1 = "AES_GCM_CTR_V1"

def init_keys(app_id, api_key, primary_key_path):
    return callBigDlFunc("float", "initKeys", app_id, api_key, primary_key_path)


def generate_encrypted_file(kms, primary_key_path, data_key_path, input_path, output_path):
    callBigDlFunc("float", "generateEncryptedFile",
                  kms, primary_key_path, data_key_path, input_path, output_path)
